dashboard/src/shared/KubeappsGrpcClient.ts

```diff
@@ -0,0 +1,81 @@
+import { grpc } from "@improbable-eng/grpc-web";
+import { PackagesServiceClientImpl } from "gen/kubeappsapis/core/packages/v1alpha1/packages";
+import {
+  GrpcWebImpl,
+  PluginsServiceClientImpl,
+} from "gen/kubeappsapis/core/plugins/v1alpha1/plugins";
+import { FluxV2PackagesServiceClientImpl } from "gen/kubeappsapis/plugins/fluxv2/packages/v1alpha1/fluxv2";
+import { HelmPackagesServiceClientImpl } from "gen/kubeappsapis/plugins/helm/packages/v1alpha1/helm";
+import { KappControllerPackagesServiceClientImpl } from "gen/kubeappsapis/plugins/kapp_controller/packages/v1alpha1/kapp_controller";
+import { Auth } from "./Auth";
+import * as URL from "./url";
+
+export class KubeappsGrpcClient {
+  private grpcWebImpl!: GrpcWebImpl;
+  private transport: grpc.TransportFactory;
+
+  // core apis
+  private packagesServiceClientImpl!: PackagesServiceClientImpl;
+  private pluginsServiceClientImpl!: PluginsServiceClientImpl;
+
+  // plugins package apis
+  private helmPackagesServiceClientImpl!: HelmPackagesServiceClientImpl;
+  private kappControllerPackagesServiceClientImpl!: KappControllerPackagesServiceClientImpl;
+  private fluxv2PackagesServiceClientImpl!: FluxV2PackagesServiceClientImpl;
+
+  constructor(transport?: grpc.TransportFactory) {
+    this.transport = transport ?? grpc.CrossBrowserHttpTransport({ withCredentials: true });
+  }
+
+  // getClientMetadata, if using token authentication, creates grpc metadata
+  // and the token in the 'authorization' field
+  public getClientMetadata() {
+    return Auth.getAuthToken()
+      ? new grpc.Metadata({ authorization: `Bearer ${Auth.getAuthToken()}` })
+      : undefined;
+  }
+
+  // getGrpcClient returns the already configured grpcWebImpl
+  // if uncreated, it is created and returned
+  public getGrpcClient = () => {
+    return (
+      this.grpcWebImpl ||
+      new GrpcWebImpl(URL.api.kubeappsapis, {
+        transport: this.transport,
+        metadata: this.getClientMetadata(),
+      })
+    );
+  };
+
+  // Core APIs
+  public getPackagesServiceClientImpl() {
+    return this.packagesServiceClientImpl || new PackagesServiceClientImpl(this.getGrpcClient());
+  }
+
+  public getPluginsServiceClientImpl() {
+    return this.pluginsServiceClientImpl || new PluginsServiceClientImpl(this.getGrpcClient());
+  }
+
+  // Plugins (packages) APIs
+  public getHelmPackagesServiceClientImpl() {
```

Comment:
Fine for now, but I'd hoped that we could query the plugin API to get a list of plugins and load the clients dynamically (even if it requires further changes to the plugin API). Not for now (it's fine as is), but do you see a way it might be possible? If so, perhaps leave a comment in the code to come back to.

Reply:
Adding a TODO there. I think it would be doable (we do have reflection things like [Reflect](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Reflect) and [Proxy](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy/Proxy)), but definitively it will require an extra (and interesting!) effort :P